src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs

```diff
@@ -0,0 +1,341 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class StreamTests
+    {
+        [Fact]
+        public static void HandleCollectionsAsync()
+        {
+            static void RunTest<TElement>()
+            {
+                foreach ((Type, int) pair in EnumerableTestData<TElement>())
+                {
+                    Type type = pair.Item1;
+                    int bufferSize = pair.Item2;
+
+                    // bufferSize * 0.9 is the threshold size from codebase, subtract 2 for [ or { characters, then create a 
+                    // string containing (threshold - 2) amount of char 'a' which when written into output buffer produces buffer 
+                    // which size equal to or very close to threshold size, then adding the string to the list, then adding a big 
+                    // object to the list which changes depth of written json and should cause buffer flush.
+                    int thresholdSize = (int)(bufferSize * 0.9 - 2);
+
+                    object obj = GetPopulatedCollection<TElement>(type, thresholdSize);
+
+                    var options = new JsonSerializerOptions
+                    {
+                        DefaultBufferSize = bufferSize,
+                        WriteIndented = true
+                    };
+
+                    var optionsWithPreservedReferenceHandling = new JsonSerializerOptions(options)
+                    {
+                        ReferenceHandling = ReferenceHandling.Preserve
+                    };
+
+                    Type elementType = typeof(TElement);
+
+                    Task[] tasks = new Task[2];
+                    tasks[0] = Task.Run(async () => await PerformSerialization(obj, type, elementType, options));
```

Comment:
So we are also testing multiple threads here (in addition to faulting on the buffer) otherwise we'd have two `await`s instead of `WaitAll`. Did you encounter issues with shared state anywhere from multiple threads?

Reply:
[As discussed offline] the tasks use separate options instances, so only faulting on the buffer is tested here. I've changed this to use two `await`s.